### PR TITLE
bz1097344 (docker-1.10 exit codes): prep work

### PR DESCRIPTION
**Pull request by edsantiago**
Background: 'docker run' exit codes will be changing in 1.10,
to differentiate between container exit status and errors from
docker itself[1]. This commit prepares for that, by
redefining the use of the docker_expected_status config
setting. Instead of 'PASS' or 'FAIL' (old values), it
is now expected to be an integer: 0 for success, X
for failure. As of this commit X is always 1, but at
some future point that will change. When it does,
the new failif_ne() helper will display the new exit
code, and it will be easy to update .ini files as
necessary.

Additionally, the mustfail() helper now requires the
expected exit status as an argument. These statuses can
actually be hardcoded in the code itself, because
mustfail() tests (once) if the installed docker binary
returns unique exit codes; if it doesn't, it changes
the expectation to 1 (what old-docker always returns).

Also includes two unrelated fixes for minor problems
discovered during this process: removed an unused
config setting, and fixed an incorrect test helper.

 [1] https://github.com/docker/docker/pull/14012/files

Note that the exit codes here are UNTESTED, and will
require tweaking once we actually have docker-1.10
available. That is expected to be minor effort.

Signed-off-by: Ed Santiago <santiago@redhat.com>

### CODE REVIEW

**Comment by edsantiago** [comment]
This is frustrating: pylint/pep8 don't issue these warnings on my system. I'll fix the comma-whitespace but will postpone global because I suspect there's a better way to preserve that state, and hope you'll guide me.

**Comment by cevich** [comment]
> This is frustrating

Likely due to differences between the Travis Ubuntu environment and yours. I lined up the version numbers best I could with RHEL 7, but it's not an exact match.  If you have any idea of how to fix this, please feel free.  I tried many times myself, but maybe there's something I missed.

Yeah, globals are generally not good to use, esp. in core code b/c you can't predict when/if test-code might be multi-process/thread and require locking.  No problem storing the data in something the caller passes in though, b/c then they can lock it if necessary.  I'll take a peek at the code and comment more...

**Comment by cevich** [comment]
Oh also, you're commit comments are lovely and well written, but way-overkill from my POV.  You're welcome to spend less time/effort on them if you wish :stuck_out_tongue_winking_eye:

File: dockertest/output.py [hunk 1]
```diff
@@ -723,7 +723,7 @@ def mustpass(cmdresult, failmsg=None):
     return cmdresult
 
 
-def mustfail(cmdresult, failmsg=None):
+def mustfail(cmdresult, expected_status, failmsg=None):
     """
```
Comment (cevich) [inline dockertest/output.py]: Might do a quick full-repo grep (or cscope) to see if there are any ``mustfail()`` callers that pass two positional-parameters, I suspect there may be either none, or a billion.  I was planning on merging this PR into the new 0.8.5 release, so the API changes are segregated.  Otherwise, just FYI -  consider when adding parameters, usually sticking them at the end as a keyword parameter is safest, and have it default to previous behaviour.

File: dockertest/output.py [hunk 2]
```diff
@@ -737,11 +737,32 @@ def mustfail(cmdresult, failmsg=None):
...
-        raise xceptions.DockerExecError("Unexpected zero exit code,"
-                                        " details: %s" % details)
+    # FIXME: temporary; remove once we no longer run on pre-1.10 docker
+    if not has_distinct_exit_codes():
+        expected_status = 1
+    if cmdresult.exit_status != expected_status:
+        raise xceptions.DockerExecError("Unexpected exit code %d; expected %d."
+                                        " Details: %s" % (cmdresult.exit_status,
+                                                          expected_status,
+                                                          details))
```
Comment (cevich) [inline dockertest/output.py]: Much better, a welcome improvement.

File: dockertest/output.py [hunk 3]
```diff
@@ -737,11 +737,32 @@ def mustfail(cmdresult, failmsg=None):
...
+    """
+    global HAS_DISTINCT_EXIT_CODES
+    if HAS_DISTINCT_EXIT_CODES is None:
+        d_run = utils.run('docker run --invalid-option nonexistent-container',
+                          ignore_status=True)
+        HAS_DISTINCT_EXIT_CODES = (d_run.exit_status > 120)
+    return HAS_DISTINCT_EXIT_CODES
 
 # This class inherits a LOT of public methods, but most of what
 # appears here is all 'behind the scenes' stuff for producing
```
Comment (cevich) [inline dockertest/output.py]: A few ideas here:
    - Make it an immutable bool-subclass, maybe singleton-pattern.  Might be overcomplicating it though.
    - Have it accept a 'state' parameter that defaults to None, then is returned as something which evaluates true/false but is mutable (e.g. ``[True]`` or ``[False]``).
    - Have ``mustfail()`` accept a callable parameter, which defaults to ``has_distinct_exit_codes``, then it just calls it instead of looking up a global.  (You'll need to flip around the function definition order).

File: dockertest/output.py [hunk 4]
```diff
@@ -737,11 +737,32 @@ def mustfail(cmdresult, failmsg=None):
...
+        raise xceptions.DockerExecError("Unexpected exit code %d; expected %d."
+                                        " Details: %s" % (cmdresult.exit_status,
+                                                          expected_status,
+                                                          details))
     return cmdresult
 
+HAS_DISTINCT_EXIT_CODES = None
+def has_distinct_exit_codes():
+    """
+    2016-03-23 **TEMPORARY** - for transition from docker-1.9 to 1.10
```
Comment (cevich) [inline dockertest/output.py]: FYI- These types of notes normally stick around /way/ longer than anyone ever intended, just look around :grinning:   These doc-strings will wind up in docker-autotest.readthedocs.org so short-and-sweet and tastefully-terse is usually the best way to go.

File: subtests/docker_cli/build/build.py [hunk 5]
```diff
@@ -225,7 +225,8 @@ def basic_postprocess(build_def, command, parameter):
             if OutputGood(build_def.dockercmd.cmdresult,
                           ignore_error=True,
                           skip=['error_check']):
-                return build_def.dockercmd.cmdresult.exit_status != 0
+                # FIXME: this is likely to fail in docker-1.10; bz1097344
+                return build_def.dockercmd.cmdresult.exit_status == 1
```
Comment (cevich) [inline subtests/docker_cli/build/build.py]: Nice, this one actually is helpful, even if it winds up being permanently-temporary (like the rest) :grinning:

File: subtests/docker_cli/dockerimport/truncated.py [hunk 6]
```diff
@@ -26,7 +26,7 @@ def cleanup(self):
         image_name = self.sub_stuff['image_name']  # assume id lookup failed
         if self.parent_subtest.config['remove_after_test']:
             dkrcmd = DockerCmd(self, 'rmi', [image_name])
-            mustfail(dkrcmd.execute())
+            mustfail(dkrcmd.execute(), 1)
```
Comment (cevich) [inline subtests/docker_cli/dockerimport/truncated.py]: Ahh okay, you did grep/cscope them all out, nice!

**Comment by cevich** [comment]
Okay, gave code a gander + comments.  This is really fantastic for a first-pass, should be really easy to sort out.  I'm going to flag it for 0.9.1 b/c API changes - there are tests which exist outside this repo. and may use some of the changed calls.  Though if you can get it to be backwards-compatible, then I can merge it into the 0.8.x series.

File: dockertest/output.py (hunk 4 above)
Comment (edsantiago) [inline dockertest/output.py]: My reasoning is that despite my best intentions, I may forget to remove this after docker-1.10. I want the next maintenance programmer to see this comment, read the datestamp, realize this code can be excised, and do so with minimal wasted time. Is there a better way to accomplish that?

**Comment by edsantiago** [comment]
Thanks for your comments, and especially on a non-work day. I need to fix that global issue, and will take a look at it on Monday.

File: dockertest/output.py [hunk 7]
```diff
@@ -171,7 +172,7 @@ def client(self):
         """
         if self._client is None:
             try:
-                return self._old_client()
+                self._client = self._old_client()
```
Comment (edsantiago) [inline dockertest/output.py]: I'm pretty sure this was meant to be memoized.

File: dockertest/output.py [hunk 8]
```diff
@@ -185,13 +186,29 @@ def server(self):
         """
         if self._server is None:
             try:
-                return self._old_server()
+                self._server = self._old_server()
```
Comment (edsantiago) [inline dockertest/output.py]: I'm pretty sure this was intended to be memoized.

File: dockertest/output_unittests.py [hunk 9]
```diff
@@ -156,7 +156,7 @@ def test_client_new(self):
         self.assertEqual(docker_version.client, '1.8.2')
         self.assertEqual(docker_version.server, '1.8.2')
 
-    def test_client_new(self):
+    def test_client_info_new(self):
```
Comment (edsantiago) [inline dockertest/output_unittests.py]: Expand the context, so you can see that this used to have the same name as the previous function. Hence one of them was masked.

**Comment by edsantiago** [comment]
> Though if you can get it to be backwards-compatible, then I can merge it into the 0.8.x series.

That's the whole idea (backward compatibility). This works with 1.8, and we have a window in which to discover what the actual docker-1.10 exit codes are and update the tests without 1.8 or 1.9 being affected.

File: dockertest/output.py (hunk 7 above)
Comment (cevich) [inline dockertest/output.py]: damnit, who wrote that!  :blush:

File: dockertest/output_unittests.py (hunk 9 above)
Comment (cevich) [inline dockertest/output_unittests.py]: Nice catch!

**Comment by cevich** [comment]
> That's the whole idea (backward compatibility). This works with 1.8, and we have a window in which to discover what the actual docker-1.10 exit codes are and update the tests without 1.8 or 1.9 being affected.

Great, okay then be sure to double-check you're not changing parameter ordering then.  Callers don't always pass keywords and there are tests that exist outside of the repository.

**Comment by edsantiago** [comment]
>there are tests that exist outside of the repository.

Ouch - really? If so, and if any of that code invokes `mustfail` with the failmsg option, it will break. Do you have a sense for how likely that is, and how much pain it would be to have the maintainers of said code update it?

**Comment by cevich** [comment]
> Ouch - really?

Yes, I have one to be sure, and I know at least the fedora guys are running Docker Autotest in some capacity.  Since it's a design feature to allow external (out-of-repo) test code to be used, that's why I keep a careful eye on API changes.  I honestly can't tell you if/how many outside tests there are, there's no requirement to publish.  It could be zero, it could be hundreds :fish:  

TBH even if we put this into 0.9.1, I control the release tagging, so we could do that tomorrow if needed :smile:   So it's not a major problem, just makes CI testing esaier b/c they can version-lock and be guaranteed the core API won't change.

Anyway, we'll figure something out.

**Comment by edsantiago** [comment]
Okay... new push includes two fixes of note: (1) backward-compatibility kludge for handling old invocations of mustfail(), and (2) use new `require_client` to test for 1.10. Also a FIXME asking if it's a good idea to do the manual exit-code check as well. Also, rebased.

File: dockertest/output.py [hunk 10]
```diff
@@ -767,25 +791,37 @@ def mustfail(cmdresult, failmsg=None):
     :type cmdresult: object convertible to string with variable exit_status
     :param failmsg: Additional messages for describing problem when cmd fails.
     """
+    # FIXME: temporary: backward compatibility for pre-20160330 code
+    # FIXME: remove before the next API-changing release
```
Comment (cevich) [inline dockertest/output.py]: oh gosh, I might need help remembering this one is here.

File: subtests/docker_cli/build/build.py [hunk 11]
```diff
@@ -225,7 +225,8 @@ def basic_postprocess(build_def, command, parameter):
             if OutputGood(build_def.dockercmd.cmdresult,
                           ignore_error=True,
                           skip=['error_check']):
-                return build_def.dockercmd.cmdresult.exit_status != 0
+                # FIXME: this is likely to fail in docker-1.10; bz1097344
```
Comment (cevich) [inline subtests/docker_cli/build/build.py]: In fact it already is, wait until you see the results from running on fedora_rawhide  :sob:

**Comment by cevich** [comment]
Wow, this is very nice work.  I'd suggest don't drive yourself completely mad by fixing up all the old tests, it's a big time-sink.  That work is much easier to do during the 0.9.1 testing when all the 'big picture' stuff comes into play.  Anyway, just a suggestion.

Is this ready to merge do you think, or was there more looking/tweaking you wanted to do?

**Comment by edsantiago** [comment]
One more commit, to fix a typo introduced in failif_ne

**Comment by edsantiago** [comment]
Remaining concerns: I haven't tested with docker-1.8, I haven't seen the rawhide results you mention, I don't have a solution to your "need help remembering" issue. Aside from those, I believe this could be merged. Subsequent problems can be addressed as needed.

**Comment by cevich** [comment]
Okay, we can test it by fire: run it through CI and see what breaks.  I'll give it a quick double-check then and merge...

File: dockertest/output.py [hunk 12]
```diff
@@ -767,25 +791,37 @@ def mustfail(cmdresult, failmsg=None):
...
+    if isinstance(expected_status, basestring):
+        failmsg = expected_status
+        expected_status = 1
+
```
Comment (cevich) [inline dockertest/output.py]: hehe, cheap hack.  Okay, I'll buy that :grinning:

File: dockertest/output.py [hunk 13]
```diff
@@ -767,25 +791,37 @@ def mustfail(cmdresult, failmsg=None):
     :type cmdresult: object convertible to string with variable exit_status
     :param failmsg: Additional messages for describing problem when cmd fails.
```
Comment (cevich) [inline dockertest/output.py]: Ahh, we need a docstring update here for ``expected_status``.  Might be good to mention there that it will be used in place of  ``failmsg`` if it's a string instead [of something else].

File: dockertest/output.py (hunk 12 above)
Comment (cevich) [inline dockertest/output.py]: Hmm, maybe this should also test if ``failmsg == None``?  Or is that being too paranoid?

**Comment by cevich** [comment]
Okay, just found one little docstring that needs some love, then this is good-to-go.  There's some possible paranoia you're welcome to fix as well, but it's probably just my ignorance.

File: dockertest/output.py (hunk 12 above)
Comment (edsantiago) [inline dockertest/output.py]: Good call. Testing for expected_status == None, though. This catches two old-style usages:

       mustfail(foo)           # missing exit status and failmsg
       mustfail(foo, "why")    # old-style call with message

File: dockertest/output.py (hunk 13 above)
Comment (edsantiago) [inline dockertest/output.py]: Good catch, thanks! New commit pushed.

**Comment by cevich** [comment]
That's why I double-check, then blindly assume everything I didn't notice isn't a problem :grinning:   Okay, merging...
